src/NuGet.Core/NuGet.DependencyResolver.Core/Remote/RemoteDependencyWalker.cs

```diff
@@ -248,34 +247,76 @@ public async Task<GraphNode<RemoteResolveResult>> WalkAsync(LibraryRange library
             return node;
         }
 
-        private Func<LibraryRange, (DependencyResult dependencyResult, LibraryDependency conflictingDependency)> ChainPredicate(Func<LibraryRange, (DependencyResult dependencyResult, LibraryDependency conflictingDependency)> predicate, GraphNode<RemoteResolveResult> node, LibraryDependency dependency)
+        /// <summary>
+        /// Walks up the package dependency graph to check for cycle, potentially degraded package versions <see cref="DependencyResult"/>.
+        /// Cycle: A -> B -> A (cycle)
+        /// Downgrade: B depends up on D 1.0. Hence this method returns a downgrade while processing D 2.0 package. 
+        /// A -> B -> C -> D 2.0 (downgrade)
+        ///        -> D 1.0
+        /// </summary>
+        /// <param name="graphEdge">Graph Edge node to check for cycle or potential degrades</param>
+        /// <param name="dependency">Transitive package dependency</param>
+        /// <param name="rootPredicate">Func delegate to invoke when processing direct package dependency</param>
+        private static (DependencyResult dependencyResult, LibraryDependency conflictingDependency) WalkParentsAndCalculateDependencyResult(
+            GraphEdge<RemoteResolveResult> graphEdge,
+            LibraryDependency dependency,
+            Func<LibraryRange, (DependencyResult dependencyResult, LibraryDependency conflictingDependency)> rootPredicate)
+        {
+            var edge = graphEdge;
+
+            //Walk up the tree starting from the grand parent upto root
+            while (edge != null)
+            {
+                var result = CalculateDependencyResult(edge.Item, edge.Edge, dependency.LibraryRange);
+
+                if (result.dependencyResult != DependencyResult.Unknown)
+                    return result;
+
+                edge = edge.OuterEdge;
+            }
+
+            return rootPredicate(dependency.LibraryRange);
+        }
+
+        private static Func<LibraryRange, (DependencyResult dependencyResult, LibraryDependency conflictingDependency)> ChainPredicate(Func<LibraryRange, (DependencyResult dependencyResult, LibraryDependency conflictingDependency)> predicate, GraphNode<RemoteResolveResult> node, LibraryDependency dependency)
```

Comment:
Consider having arguments on new lines to improve readability.

Reply:
Fixed in the latest commit.